File: c++/gfg/MergeSort.cpp
#include <iostream>

using namespace std;


void merge(int arr[], int n, int low, int mid, int high) {
    int temp_arr[high-low+1];
    int i = low, j = mid+1, k = 0;
    while (i <= mid && j <= high)
    {
        if (arr[i] < arr[j]) {
            temp_arr[k] = arr[i];
            i++;
        } else {
            temp_arr[k] = arr[j];
            j++;
        }
        k++;
    }
    while (i <= mid)
    {
        temp_arr[k] = arr[i];
        i++; k++;
    }
    while (j <= mid)
    {
        temp_arr[k] = arr[j];
        j++; k++;
    }
    for (int i=low; i<=high; i++)
    {
        arr[i] = temp_arr[i-low];
    }
}

void merge_sort(int arr[], int n, int low, int high)
{
    if (low <= high)
    {
        if (low == high)
            return;
        int mid = low + (high-low)/2;
        merge_sort(arr, n, low, mid);
        merge_sort(arr, n, mid+1, high);
        merge(arr, n, low, mid, high);
        return;
    }

}

int main()
{
    int n;
    cin >> n;
    int arr[n];
    for (int i=0; i<n; i++) cin >> arr[i];
    merge_sort(arr, n, 0, n-1);
    for (int i=0; i<n; i++) cout << arr[i] << " ";
    cout << "\n";
}

File: python/Algo/mergeSort.py

def merge(arr, low, mid, high):
    tempArr = [0] * (high-low+1)
    i, j, k = low, mid+1, 0
    while i <= mid and j <= high:
        if arr[i] <= arr[j]:
            tempArr[k] = arr[i]
            i += 1
        else:
            tempArr[k] = j
            j += 1
        k += 1
    while i <= mid:
        tempArr[k] = arr[i]
        i += 1
        k += 1
    while j <= high:
        tempArr[k] = arr[j]
        j += 1
        k += 1
    for i in range(low, high+1):
        arr[i] = tempArr[i-low]


def mergeSort(arr, low, high):
    if high == low:
        return
    mid = low + ((high - low) >> 1)
    mergeSort(arr, low, mid)
    mergeSort(arr, mid+1, high)
    merge(arr, low, mid, high)
    print(arr)


arr = [8, 3, 7, 8, 3, 4, 79, 2, 4, 73, 6, 3, 78, 34]
mergeSort(arr, 0, len(arr)-1)
print(arr)
